### DISCUSSION TITLE
Could r-universe replace retired CRAN time machine?

### DISCUSSION THREAD

**Discussion by skranz** [r-universe-org/help#225] [Ideas]
Dear R-universe developers,

first thanks for the great infrastructure you developed and maintain. I just wanted to float an idea (related to https://github.com/r-universe-org/help/discussions/91) after reading that the Microsoft CRAN time machine will retire (https://blog.revolutionanalytics.com/2023/01/mran-time-machine-retired.html). 

The retirement of the CRAN time machine is a pity, as having old snapshots of CRAN including binaries can be quite useful, e.g. to keep old code and data supplements from research articles reproducible, or to quickly install a package that was removed from CRAN. I wonder whether you could imagine that the r-universe framework could offer a replacement for the CRAN time machine. Daily snapshots as on the CRAN time machine are probably quite resource-intensive, but already monthly or even yearly snapshots might be helpful.

That being said, I have no clue how much time and effort it would cost to implement such a replacement with the r-universe framework.

- Does it suffice to generate a Github user like "cran-snapshot-2019-01-01" for every snapshot date, write a script that copies the package sources from CRAN or a particular old CRAN time machine snapshot to correspondingly generated Github repositiories of that user and then use the standard r-universe infrastructure to compile them? That might be similar to the scripts, you use to scrap repositiories from CRAN, but perhaps it is considerably more complicated.

- I guess one problem is that binaries of older CRAN package snapshots probably should be available for corresponding older R versions, so probably one would kind of need an extension to r-universe that allows to specify that binaries should be generated for particular R versions... that might be complicated.

- There might be other problems, I don't see.

As said, I just wanted to float the idea...

**Comment by arnim** [comment]
CC @chainsawriot

**Reply by arnim** [reply]
https://github.com/chainsawriot/rang @skranz

**Comment by dustin-34** [comment]
Hey! Dustin here, and we have a team of experienced professionals ready to assist you by taking your [take my course online for me](https://acemyonlineclasses.com/online-courses.php)  Whether it’s attending lectures, participating in discussions, or completing quizzes and exams, we’ve got you covered. Focus on the subjects that matter most to you while we handle the rest.

**Comment by eitsupi** [comment]
Maybe https://r-multiverse.org/ 's production repo.

**Reply by llrs** [reply]
Isn't r-multiverse only snapshotting since 2024? How would a user trying to get CRAN version from 2005-02-15 work with r-multiverse?

**Comment by jeroen** [comment]
P3M hosts public snapshots of CRAN, for example you can use:

```r
options(repos = c(CRAN = "https://p3m.dev/cran/2025-01-15"))
```

In R-universe we use this such that you can set an option in your `config.json` such that all packages are built and checked against a given snapshot, for example https://github.com/r-multiverse/staging/blob/main/config.json. This way you can maintain a set of R packages that is tested to work with given versions of the dependenciest

**Reply by maelle** [reply]
Docs link for what @jeroen says: https://docs.r-universe.dev/publish/troubleshoot-build.html#configure-some-options-with-config.json

**Comment by maelle** [comment]
You could also download your own snapshot from R-universe: https://docs.r-universe.dev/install/reproducibility.html#snapshots
